biz.aQute.bndlib/src/aQute/bnd/cdi/AnnotationReader.java

```diff
@@ -0,0 +1,340 @@
+					TypeRef typeRef = analyzer.getTypeRef(returnType.binary);
+					BeanDef producer = new BeanDef();
+					producer.service.add(typeRef);
+					definitions.add(producer);
+				}
+				break;
+			case TYPE_USE :
+				if (definitions.get(0).serviceOrigin != null && definitions.get(0).serviceOrigin == ElementType.TYPE) {
+					analyzer.error("@Service cannot be used both on TYPE and TYPE_USE on class %s", clazz);
+					break;
+				}
+				definitions.get(0).serviceOrigin = ElementType.TYPE_USE;
+				int targetIndex = annotation.getTargetIndex();
+				if (targetIndex == Annotation.TARGET_INDEX_EXTENDS) {
+					definitions.get(0).service.add(extendsClass);
+				} else {
+					definitions.get(0).service.add(interfaces[targetIndex]);
+				}
+				break;
+			case TYPE :
+				if (definitions.get(0).serviceOrigin != null
+					&& definitions.get(0).serviceOrigin == ElementType.TYPE_USE) {
+					analyzer.error("@Service cannot be used both on TYPE and TYPE_USE on class %s", clazz);
+					break;
+				}
+				definitions.get(0).serviceOrigin = ElementType.TYPE;
+				Object[] serviceClasses = annotation.get("value");
+
+				if (serviceClasses != null && serviceClasses.length > 0) {
+					for (Object serviceClass : serviceClasses) {
+						definitions.get(0).service.add((TypeRef) serviceClass);
+					}
+				} else if (interfaces != null && interfaces.length > 0) {
+					for (TypeRef inter : interfaces) {
+						definitions.get(0).service.add(inter);
+					}
+				} else {
+					definitions.get(0).service.add(clazz.getClassName());
+				}
+				break;
+			default :
+				// this can't happen, @Service is limited to the above targets
+		}
+	}
+
+	@Override
+	public void classBegin(int access, TypeRef name) {
+		definitions.get(0).implementation = name;
+
+		PackageDef packageDef = packageInfos.computeIfAbsent(name.getPackageRef(), k -> {
+			Clazz packageInfoClazz = analyzer.getPackageInfo(k);
+			if (packageInfoClazz != null) {
+				try {
+					PackageDef pd = new PackageDef();
+					packageInfoClazz.parseClassFileWithCollector(pd);
+					return pd;
+				} catch (Exception e) {
+					analyzer.exception(e, "Error while processing package-info of class %s", clazz);
+				}
+			}
+			return new PackageDef();
+		});
+
+		if (packageDef.marked != null) {
+			definitions.get(0).marked = packageDef.marked.matches(name.getFQN());
+		}
+	}
+
+	@Override
+	public void classEnd() throws Exception {
+		member = null;
+		referenceDef = null;
+	}
+
+	@Override
+	public void extendsClass(TypeRef name) {
+		this.extendsClass = name;
+	}
+
+	@Override
+	public void field(FieldDef field) {
+		this.member = field;
+	}
+
+	@Override
+	public void implementsInterfaces(TypeRef[] interfaces) {
+		this.interfaces = interfaces;
+	}
+
+	@Override
+	public void memberEnd() {
+		member = null;
+		referenceDef = null;
+	}
+
+	@Override
+	public void method(MethodDef method) {
+		if (method.isAbstract() || method.isBridge())
+			return;
```

Comment:
if you early return, you should probably set `member` to null so that an annotation on this rejected method is not processed on the stale value of `member`.